server/src/main/java/de/uftos/repositories/solver/timefold/constraints/ConstraintDefinitionFactory.java

```diff
@@ -0,0 +1,1073 @@
+            default -> throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      case "tags" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          switch (resource.getResourceType()) {
+            case TIMETABLE ->
+                resultSet.addAll(((TimetableSolutionTimefoldInstance) resource).getTags());
+            case STUDENT ->
+                resultSet.addAll(((StudentTimefoldInstance) resource).getProvidedTagsList());
+            case TEACHER ->
+                resultSet.addAll(((TeacherTimefoldInstance) resource).getProvidedTagsList());
+            case STUDENT_GROUP ->
+                resultSet.addAll(((StudentGroupTimefoldInstance) resource).getProvidedTagsList());
+            case ROOM -> resultSet.addAll(((RoomTimefoldInstance) resource).getProvidedTagsList());
+            case SUBJECT ->
+                resultSet.addAll(((SubjectTimefoldInstance) resource).getProvidedTagsList());
+            case GRADE ->
+                resultSet.addAll(((GradeTimefoldInstance) resource).getProvidedTagsList());
+            case TIMESLOT ->
+                resultSet.addAll(((TimeslotTimefoldInstance) resource).getProvidedTagsList());
+            default -> throw new IllegalStateException();
+          }
+        }
+        return resultSet;
+      };
+      default -> throw new IllegalStateException();
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> convertElement(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() == UcdlToken.NUMBER || ast.getToken() == UcdlToken.SIZE) {
+      Function<List<ResourceTimefoldInstance>, Number> function = convertNumberElement(ast, params);
+      return (parameters) -> (ResourceTimefoldInstance) function.apply(parameters);
+    }
+
+    if (ast.getToken() != UcdlToken.ELEMENT) {
+      throw new IllegalStateException();
+    }
+    ElementDto element = (ElementDto) ast;
+
+    Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> name =
+        convertValueReference(element.name(), params);
+
+    List<Function<
+        List<ResourceTimefoldInstance>,
+        Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+        >> attributes = new ArrayList<>();
+
+    for (AbstractSyntaxTreeDto attribute : element.attributes()) {
+      attributes.add(convertAttribute(attribute, params));
+    }
+
+    return (parameters) -> {
+      Set<ResourceTimefoldInstance> elementReference = new HashSet<>();
+      elementReference.add(name.apply(parameters));
+
+      for (Function<
+            List<ResourceTimefoldInstance>,
+            Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+            > function : attributes) {
```

Comment:
Those types are killing me lmao

Reply:
You mean the generics?
Yeah... They are fun...